### ISSUE TITLE
Typescript compiler reports errors in react-native project

### ISSUE THREAD

**Issue by curthipster** [nrwl/nx#12417]
The Typescript compiler fails in a newly generated react-native project. I'd like to be able to run the compiler locally and as part of CI to detect errors.

## Current Behavior
Running the typescript compiler in the mobile project results in 4 configuration errors and it should detect real issues in the code.

## Expected Behavior
The typescript compiler should work out-of-the-box for a newly generated project.

## Steps to Reproduce
1. Create a fresh workspace with a react native app: `npx create-nx-workspace@latest happynrwl --preset=react-native --appName=mobile`
2. `cd happynrwl/apps/mobile`
3. `npx tsc --noEmit`
4. Observe 4 errors in tsconfig.json

### Failure Logs
```bash
npx tsc --noEmit
tsconfig.json:15:5 - error TS6306: Referenced project '/Users/chrispruett/dev/nx-rn-oct4/apps/mobile/tsconfig.app.json' must have setting "composite": true.

15     {
       ~
16       "path": "./tsconfig.app.json"
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
17     },
   ~~~~~

tsconfig.json:15:5 - error TS6310: Referenced project '/Users/chrispruett/dev/nx-rn-oct4/apps/mobile/tsconfig.app.json' may not disable emit.

15     {
       ~
16       "path": "./tsconfig.app.json"
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
17     },
   ~~~~~

tsconfig.json:18:5 - error TS6306: Referenced project '/Users/chrispruett/dev/nx-rn-oct4/apps/mobile/tsconfig.spec.json' must have setting "composite": true.

18     {
       ~
19       "path": "./tsconfig.spec.json"
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
20     }
   ~~~~~

tsconfig.json:18:5 - error TS6310: Referenced project '/Users/chrispruett/dev/nx-rn-oct4/apps/mobile/tsconfig.spec.json' may not disable emit.

18     {
       ~
19       "path": "./tsconfig.spec.json"
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
20     }
   ~~~~~


Found 4 errors in the same file, starting at: tsconfig.json:15
```

### Environment
```
$ nx report

 >  NX   Report complete - copy this into the issue template

   Node : 16.17.1
   OS   : darwin arm64
   npm  : 8.15.0
   
   nx : 14.8.3
   @nrwl/angular : Not Found
   @nrwl/cypress : 14.8.3
   @nrwl/detox : 14.8.3
   @nrwl/devkit : 14.8.3
   @nrwl/esbuild : Not Found
   @nrwl/eslint-plugin-nx : 14.8.3
   @nrwl/expo : Not Found
   @nrwl/express : Not Found
   @nrwl/jest : 14.8.3
   @nrwl/js : 14.8.3
   @nrwl/linter : 14.8.3
   @nrwl/nest : Not Found
   @nrwl/next : Not Found
   @nrwl/node : Not Found
   @nrwl/nx-cloud : Not Found
   @nrwl/nx-plugin : Not Found
   @nrwl/react : 14.8.3
   @nrwl/react-native : 14.8.3
   @nrwl/rollup : 14.8.3
   @nrwl/schematics : Not Found
   @nrwl/storybook : 14.8.3
   @nrwl/web : 14.8.3
   @nrwl/webpack : 14.8.3
   @nrwl/workspace : 14.8.3
   typescript : 4.8.4
   ---------------------------------------
   Local workspace plugins:
   ---------------------------------------
   Community plugins:
```

## Additional Context and Troubleshooting

It's easy to make those errors go away by following the instructions in the output. Specifically, add the following `"compilerOptions"` to each file:

```javascript
    "composite": true,
    "noEmit": false
```

However, I believe that there are deeper Typescript configuration issues, and I'm not convinced that's the right fix. If you do that and then introduce a type error in the code, for example:

Change line 28 of `src/app/App.tsx` to:

```typescript
  const [whatsNextYCoord, setWhatsNextYCoord] = useState<number>('this must be a number');
```

Xcode properly reports a typescript error

![Screen Shot 2022-10-04 at 5 30 09 PM](https://user-images.githubusercontent.com/577293/193954476-425491a0-2a9d-4fc6-97ff-9df18968bf8a.png)

However, the command line compiler reports no issues:

```bash
$ npx tsc --noEmit
$ echo $?
0
```

**Comment by Miikis** [comment]
I encountered this on a newly generated expo project as well. For me, setting `compilerOptions.declaration` to `true` (in addition to setting `compilerOptions.composite` to `true` _and_ `compilerOptions.noEmit` to `false`) seems to get the typechecking working again. 

Also unsure if this is the best resolution tho.

**Comment by curthipster** [comment]
@Miikis Adding that setting makes the warnings go away, but it's not reporting type errors when running the typescript compiler.

**Comment by curthipster** [comment]
I realized that this is a composite typescript project, so I have specify one of leaf tsconfig files (h/t to @Miikis for helping me understand this).

To typecheck the app: `npx tsc --noEmit -p apps/mobile/tsconfig.app.json`

To typecheck the tests: `npx tsc --noEmit -p apps/mobile/tsconfig.spec.json`

With that knowledge, I was able to setup executors in the project:

```
    "typecheck": {
      "executor": "@nrwl/js:tsc",
      "outputs": ["{options.outputPath}"],
      "options": {
        "outputPath": "dist/apps/mobile",
        "main": "apps/mobile/src/main.tsx",
        "tsConfig": "apps/mobile/tsconfig.app.json",
        "rootDir": "../.."
      }
    },

    "typecheck-tests": {
      "executor": "@nrwl/js:tsc",
      "outputs": ["{options.outputPath}"],
      "options": {
        "outputPath": "dist/apps/mobile",
        "main": "apps/mobile/src/main.tsx",
        "tsConfig": "apps/mobile/tsconfig.spec.json",
        "rootDir": "../.."
      }
    },
```

For that to work, I had to make a couple of changes to the tsconfig. I'm sharing them here because I'm not certain whether there's a bug in the project template or if I'm doing something wrong.

In `apps/mobile/tsconfig.json`:
```
    "noEmit": false,
    "declaration": true,
```

In `apps/mobile/tsconfig.app.json` and `apps/mobile/tsconfig.spec.json`:
```
    "composite": true
```

**Comment by Miikis** [comment]
@devor there's two separate issues here. One has to do with trying to typecheck a composite project without building it; the other (at least in my case) has to do with incorrect tsconfig compiler options for a composite project. 

At least for me, the latter case was due to some defaults I added to the tsconfig.base.json. Adding the correct compiler-options overrides in the target project's tsconfig resolved that issue for me. (See above comments for the necessary options changes).
